File: Java/Exercises/BrainTrainer/app/src/main/java/com/example/braintrainer/MainActivity.java
package com.example.braintrainer;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Random;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/** @Autor Created by Anubys on the 26.03.2020 */


public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();

    @Nullable
    @BindView(R.id.tv_timer)
    TextView textViewTimer;
    @Nullable
    @BindView(R.id.tv_task)
    TextView textViewTask;
    @Nullable
    @BindView(R.id.tv_interim_result)
    TextView textViewInterimResult;
    @Nullable
    @BindView(R.id.tv_selection1)
    TextView textViewSelection1;
    @Nullable
    @BindView(R.id.tv_selection2)
    TextView textViewSelection2;
    @Nullable
    @BindView(R.id.tv_selection3)
    TextView textViewSelection3;
    @Nullable
    @BindView(R.id.tv_selection4)
    TextView textViewSelection4;
    @Nullable
    @BindView(R.id.tv_score)
    TextView textViewScore;
    @Nullable
    @BindView(R.id.btn_play_again)
    Button buttonPlayAgain;

    private ArrayList<Integer> answers = new ArrayList<>();
    private Unbinder unbinder;
    private CountDownTimer countDownTimer;
    private int locationOfCorrectAnswer;
    private int score = 0;
    private int numberOfQuestions = 0;


    //* ************************************************ *
    //*               L I F E - C Y C L E                *
    //* ************************************************ *
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "TAG - MainActivity - onCreate()");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        unbinder = ButterKnife.bind(this);

        setListener();
    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "TAG - MainActivity - onDestroy()");
        super.onDestroy();

        if (unbinder != null) {
            unbinder.unbind();
        }

        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
    }


    //* ************************************************ *
    //*         H E L P E R  -  M E T H O D S            *
    //* ************************************************ *
    private void setListener() {
        Log.d(TAG, "TAG - MainActivity - setListener()");

        if (buttonPlayAgain != null) {
            buttonPlayAgain.setOnClickListener(view -> playAgain());
        }

        if (textViewSelection1 != null) {
            textViewSelection1.setOnClickListener(this::chooseAnswer);
        }

        if (textViewSelection2 != null) {
            textViewSelection2.setOnClickListener(this::chooseAnswer);
        }

        if (textViewSelection3 != null) {
            textViewSelection3.setOnClickListener(this::chooseAnswer);
        }

        if (textViewSelection4 != null) {
            textViewSelection4.setOnClickListener(this::chooseAnswer);
        }
    }

    private void playAgain() {
        Log.d(TAG, "TAG - MainActivity - playAgain()");
        score = 0;
        numberOfQuestions = 0;

        if (textViewScore != null) {
            String sScore = score + getResources().getString(R.string.txt_slash) + numberOfQuestions;
            textViewScore.setText(sScore);
        }

        if (textViewTimer != null) {
            textViewTimer.setText(getResources().getString(R.string.txt_30_secunden));
        }

        if (buttonPlayAgain != null) {
            buttonPlayAgain.setVisibility(View.INVISIBLE);
        }

        if (textViewScore != null) {
            textViewScore.setText(getResources().getString(R.string.txt_0_0));
        }

        newQuestion();
        startCountDownTimer();
    }

    public void newQuestion() {
        Log.d(TAG, "TAG - MainActivity - newQuestion()");
        Random rand = new Random();

        int a = rand.nextInt(21);
        int b = rand.nextInt(21);

        if (textViewTask != null) {
            String sTask = a + getResources().getString(R.string.txt_plus) + b;
            textViewTask.setText(sTask);
        }

        locationOfCorrectAnswer = rand.nextInt(4);
        answers.clear();

        for (int i = 0; i < 4; i++) {
            if (i == locationOfCorrectAnswer) {
                answers.add(a + b);
            } else {
                int wrongAnswer = rand.nextInt(41);
                while (wrongAnswer == a + b) {
                    wrongAnswer = rand.nextInt(41);
                }

                answers.add(wrongAnswer);
            }
        }

        if (textViewSelection1 != null && textViewSelection2 != null && textViewSelection3 != null && textViewSelection4 != null) {
            textViewSelection1.setText(getResources().getString(R.string.txt_leer, answers.get(0)));
            textViewSelection2.setText(getResources().getString(R.string.txt_leer, answers.get(1)));
            textViewSelection3.setText(getResources().getString(R.string.txt_leer, answers.get(2)));
            textViewSelection4.setText(getResources().getString(R.string.txt_leer, answers.get(3)));
        }
    }

    private void startCountDownTimer() {
        Log.d(TAG, "TAG - MainActivity - startCountDownTimer()");
        countDownTimer = new CountDownTimer(30000, 1000) {
            @Override
            public void onTick(long l) {
                if (textViewTimer != null) {
                    String sTimer = l / 1000 + getResources().getString(R.string.txt_s);
                    textViewTimer.setText(sTimer);
                }
            }

            @Override
            public void onFinish() {
                if (textViewInterimResult != null) {
                    textViewInterimResult.setText(getResources().getString(R.string.txt_done));
                }

                if (buttonPlayAgain != null) {
                    buttonPlayAgain.setVisibility(View.VISIBLE);
                }
            }
        }.start();
    }

    public void chooseAnswer(View view) {
        Log.d(TAG, "TAG - MainActivity - chooseAnswer()");
        if (Integer.toString(locationOfCorrectAnswer).equals(view.getTag().toString())) {
            if (textViewInterimResult != null) {
                textViewInterimResult.setText(getResources().getString(R.string.txt_correct));
            }

            score++;
        } else {
            if (textViewInterimResult != null) {
                textViewInterimResult.setText(getResources().getString(R.string.txt_wrong));
            }
        }

        numberOfQuestions++;
        if (textViewScore != null) {
            String sScore = score + getResources().getString(R.string.txt_slash) + numberOfQuestions;
            textViewScore.setText(sScore);
        }

        newQuestion();
    }
}

File: Kotlin/Examples/ExpandableRecyclerViewDemo/app/src/main/java/com/anubys/example/expandablerecyclerviewdemo/MainActivity.kt
package com.anubys.example.expandablerecyclerviewdemo

/** @Author Created by Anubys on the 04.12.2020 */

import android.os.Bundle
import android.util.Log

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager

import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity(){
    private val tag = MainActivity::class.java.simpleName

    private val itemAdapter by lazy {
        DummyDataAdapter()
    }


    //* ************************************************ *
    //*               L I F E - C Y C L E                *
    //* ************************************************ *
    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(tag, "TAG - MainActivity - onCreate()")
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initRecyclerView()
    }


    //* ************************************************ *
    //*         H E L P E R  -  M E T H O D S            *
    //* ************************************************ *
    private fun initRecyclerView() {
        Log.d(tag, "TAG - MainActivity - initRecyclerView()")

        itemAdapter.setItem(getDummyDataToPass())
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = itemAdapter
        recyclerView.setHasFixedSize(true)
    }

    private fun getDummyDataToPass(): ArrayList<DummyParentDataItem> {
        Log.d(tag, "TAG - MainActivity - getDummyDataToPass()")

        val arrDummyData: ArrayList<DummyParentDataItem> = ArrayList()

        var childDataItems: ArrayList<DummyChildDataItem> = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("A Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("A Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("A Child 3")))
        arrDummyData.add(DummyParentDataItem(ItemParent("A Parent, 3 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("B Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("B Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("B Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("B Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("B Child 5")))
        childDataItems.add(DummyChildDataItem(ItemChild("B Child 6")))
        arrDummyData.add(DummyParentDataItem(ItemParent("B Parent, 6 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 5")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 6")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 7")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 8")))
        childDataItems.add(DummyChildDataItem(ItemChild("C Child 9")))
        arrDummyData.add(DummyParentDataItem(ItemParent("C Parent, 9 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("D Child 1")))
        arrDummyData.add(DummyParentDataItem(ItemParent("D Parent, 1 Children", childDataItems)))

        childDataItems = ArrayList()
        arrDummyData.add(DummyParentDataItem(ItemParent("E Parent, 0 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("F Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("F Child 2")))
        arrDummyData.add(DummyParentDataItem(ItemParent("F Parent, 2 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("G Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("G Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("G Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("G Child 4")))
        arrDummyData.add(DummyParentDataItem(ItemParent("G Parent, 4 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 5")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 6")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 7")))
        childDataItems.add(DummyChildDataItem(ItemChild("H Child 8")))
        arrDummyData.add(DummyParentDataItem(ItemParent("H Parent, 8 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("I Child 1")))
        arrDummyData.add(DummyParentDataItem(ItemParent("I Parent, 1 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("J Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("J Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("J Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("J Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("J Child 5")))
        arrDummyData.add(DummyParentDataItem(ItemParent("J Parent, 5 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 5")))
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 6")))
        childDataItems.add(DummyChildDataItem(ItemChild("K Child 7")))
        arrDummyData.add(DummyParentDataItem(ItemParent("K Parent, 7 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("L Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("L Child 2")))
        arrDummyData.add(DummyParentDataItem(ItemParent("L Parent, 2 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("M Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("M Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("M Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("M Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("M Child 5")))
        arrDummyData.add(DummyParentDataItem(ItemParent("M Parent, 5 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 2")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 3")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 4")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 5")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 6")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 7")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 8")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 9")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 10")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 11")))
        childDataItems.add(DummyChildDataItem(ItemChild("N Child 12")))
        arrDummyData.add(DummyParentDataItem(ItemParent("N Parent, 12 Children", childDataItems)))

        childDataItems = ArrayList()
        childDataItems.add(DummyChildDataItem(ItemChild("O Child 1")))
        childDataItems.add(DummyChildDataItem(ItemChild("O Child 2")))
        arrDummyData.add(DummyParentDataItem(ItemParent("O Parent, 2 Children", childDataItems)))

        return (arrDummyData)
    }
}
